### PR TITLE
MLv2 filter schema

### PR DESCRIPTION
**Pull request by metamben**
Malli schema for filter expressions part of #28692.

### CODE REVIEW

File: src/metabase/lib/schema/expression.cljc [hunk 1]
```diff
@@ -49,9 +49,10 @@
 
 ;;; Any type of expression that can appear in an `:=` or `!=`. I guess this is currently everything?
 (mr/def ::equality-comparable
-  [:or
-   ::boolean
-   ::string
-   ::number
-   ::temporal
-   ::ref/ref])
+  [:maybe
```
Comment (metamben) [inline src/metabase/lib/schema/expression.cljc:52]: Make `nil` equality-comparable.

File: src/metabase/lib/schema/filter.cljc [hunk 2]
```diff
@@ -3,15 +3,116 @@
...
+   [:lon-field orderable]
+   [:lat-max   orderable]
+   [:lon-min   orderable]
+   [:lat-min   orderable]
+   [:lon-max   orderable]])
+
+;; [:= ... nil], [:!= ... nil], [:or [:= ... nil] [:= ... ""]], [:and [:!= ... nil] [:!= ... ""]]
+(doseq [op [::is-null ::not-null ::is-empty ::not-empty]]
+  (defclause op
+    ::ref/field))
```
Comment (metamben) [inline src/metabase/lib/schema/filter.cljc:56]: Should these be allowed for all/string expressions?

File: src/metabase/lib/schema/filter.cljc [hunk 3]
```diff
@@ -3,15 +3,116 @@
...
+
+(def ^:private string
+  [:schema [:ref :metabase.lib.schema.expression/string]])
+
+;; [:does-not-contain ...] = [:not [:contains ...]]
+(doseq [op [::starts-with ::ends-with ::contains ::does-not-contain]]
+  (mr/def op
+    [:catn
+     [:clause [:= (keyword (name op))]]
+     [:options [:merge ::common/options string-filter-options]]
```
Comment (metamben) [inline src/metabase/lib/schema/filter.cljc:69]: Migrated options to the standard position.

File: src/metabase/lib/schema/filter.cljc [hunk 4]
```diff
@@ -3,15 +3,116 @@
...
+           [:n [:or :int [:enum :current :last :next]]]
+           [:unit relative-datetime-unit]]]])
+
+(defclause ::segment
+  [:catn [:segment-id [:or ::common/int-greater-than-zero ::common/non-blank-string]]])
+
+(defclause ::case
+  [:+ [:catn
+       [:pred [:schema [:ref ::filter]]]
+       [:expr [:schema [:ref :metabase.lib.schema.expression/boolean]]]]])
```
Comment (metamben) [inline src/metabase/lib/schema/filter.cljc:94]: `:case` expression with boolean results. I wonder if we'll run into difficulties when `:case` expressions for other types are added.

File: src/metabase/util/malli/registry.cljc [hunk 5]
```diff
@@ -2,11 +2,12 @@
...
   (:require
    [malli.core :as mc]
-   [malli.registry :as mr])
+   [malli.registry :as mr]
+   [malli.util :as mut])
   #?(:cljs (:require-macros [metabase.util.malli.registry])))
 
 (defonce ^:private registry*
-  (atom (mc/default-schemas)))
+  (atom (merge (mc/default-schemas) (mut/schemas))))
```
Comment (metamben) [inline src/metabase/util/malli/registry.cljc:10]: `(mut/schemas)` add support for declarative merge, untion, etc.

File: test/metabase/lib/schema/filter_test.cljc [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+         [:contains {:case-sensitive false} "abc" "a"]
+         [:does-not-contain "abc" "a"]
+         [:does-not-contain {:case-sensitive false} "abc" "a"]
+         [:time-interval field :last :hour]
+         [:time-interval field 4 :hour]
+         [:time-interval {:include-current true} field :next :default]
+         [:segment 1]
+         [:segment "segment-id"]
+         [:case [:= 1 1] true [:not-null field] [:< 0 1]]]]
+    (is (= (known-filter-ops) (filter-ops filter-expr)))
```
Comment (metamben) [inline test/metabase/lib/schema/filter_test.cljc:82]: Make sure that all filter clauses are covered.

File: src/metabase/util/malli/registry.cljc (hunk 5 above)
Comment (bshepherdson) [reply src/metabase/util/malli/registry.cljc:10]: Fancy!

**Review by bshepherdson — APPROVED** [review]

File: src/metabase/lib/schema/filter.cljc [hunk 7]
```diff
@@ -3,15 +3,116 @@
...
+
+(doseq [op [::and ::or]]
+  (defclause op [:repeat {:min 2} [:schema [:ref ::filter]]]))
+
+(defclause ::not
+  [:schema [:ref ::filter]])
+
+(doseq [op [::= ::!=]]
+  (defclause op
+    [:repeat {:min 2} eq-comparable]))
```
[the remark this answers is not in the document]
Comment (metamben) [reply src/metabase/lib/schema/filter.cljc:34]: Yes, `mr/def` doesn't define variables. BTW, the tests pass in CLJS, isn't that enough of a proof?

File: test/metabase/lib/schema/filter_test.cljc (hunk 6 above)
Comment (snoe) [reply test/metabase/lib/schema/filter_test.cljc:82]: Is this an open schema or could we test that it disallows unknown filter ops?

Comment (metamben) [reply test/metabase/lib/schema/filter_test.cljc:82]: I think it makes sense to close the schema. Is that what you are suggesting?

File: src/metabase/lib/schema/filter.cljc [hunk 8]
```diff
@@ -3,15 +3,116 @@
...
+  [:catn [:segment-id [:or ::common/int-greater-than-zero ::common/non-blank-string]]])
+
+(defclause ::case
+  [:+ [:catn
+       [:pred [:schema [:ref ::filter]]]
+       [:expr [:schema [:ref :metabase.lib.schema.expression/boolean]]]]])
 
+;; Boolean literals are not included here because they are not very portable
+;; across different databases. In places where they should also be allowed
+;; the :metabase.lib.schema.expression/boolean schema can be used.
```
Comment (camsaul) [inline src/metabase/lib/schema/filter.cljc:98]: While it's true that some databases don't have the equivalent of `true` or `false`, we have `metabase.driver.sql.query-processor` methods to work around that and convert them to the appropriate representation for the current database e.g. `1` or `0`, so I think it's fine to support boolean literals in general, it should still work ok

File: test/metabase/lib/schema/filter_test.cljc [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+                 (not (#{:field} (first top))))
+          (recur (into others (rest top)) (conj ops (first top)))
+          (recur others ops)))
+      ops)))
+
+(deftest ^:parallel filter-test
+  (testing "valid filters"
+    (let [field [:field 1 {:lib/uuid (str (random-uuid))}]
+          boolean-field [:field 2 {:lib/uuid (str (random-uuid))
+                                   :base-type :type/Boolean}]
```
Comment (camsaul) [inline test/metabase/lib/schema/filter_test.cljc:56]: Just a heads up, I moved the `:field` options to the second arg in #29010. If this PR lands first then I can update these tests

**Review by camsaul — APPROVED** [review]

File: src/metabase/lib/schema/filter.cljc (hunk 2 above)
Comment (camsaul) [reply src/metabase/lib/schema/filter.cljc:56]: I think `:is-null` and `:not-null` are ok for everything, so we should allow it for any expression.

I don't `:is-empty` or `:not-empty` make sense for everything. For example don't think it makes sense to ask whether a DateTime is empty. But any sort of `::string` expression, including `:type/Text` refs, should be ok

**Pull request merged**